### PR TITLE
SLCORE-139 Get notifications about events in SonarQube

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+  /**
+   * Get all notification events for a set of projects after a given timestamp. 
+   * Returns an empty list if an error occurred making the request or parsing the response.
+   */
+  @CheckForNull
+  public List<SonarQubeNotification> request(Map<String, ZonedDateTime> requestConfig) {
+    String path = getPath(requestConfig);
+    WsResponse wsResponse = wsClient.rawGet(path);
+    if (!wsResponse.isSuccessful()) {
+      LOG.debug("Failed to get notifications: {}, {}", wsResponse.code(), wsResponse.content());
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java]: MAJOR Invoke method(s) only conditionally. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    builder.append("&from=");
+    builder.append(requestConfig.values().stream()
+      .map(timestamp -> timestamp.format(TIME_FORMATTER))
+      .map(StringUtils::urlEncode)
+      .collect(Collectors.joining(",")));
+    return builder.toString();
+  }
+
+  private List<SonarQubeNotification> parseResponse(String contents) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java]: MAJOR Make "parseResponse" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+      for (JsonElement el : events) {
+        JsonObject event = el.getAsJsonObject();
+        String category = event.get("category").getAsString();
+        String message = event.get("message").getAsString();
+        String link = event.get("link").getAsString();
+
+        JsonElement projectEl = event.get("project");
+        String projectKey = (projectEl != null) ? projectEl.getAsString() : null;
+        // TODO
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    this.timer = timer;
+    this.task = task;
+    this.timer.scheduleAtFixedRate(task, randomizeInitialDelay(), DELAY);
+  }
+
+  /**
+   * Generated a random number between DELAY*0.5 and DELAY*1.5
+   */
+  private static int randomizeInitialDelay() {
+    Random r = new Random();
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java]: CRITICAL Use a cryptographically strong random number generator (RNG) like "java.security.SecureRandom" in place of this PRNG [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2245)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+      .map(StringUtils::urlEncode)
+      .collect(Collectors.joining(",")));
+
+    builder.append("&from=");
+    builder.append(requestConfig.values().stream()
+      .map(timestamp -> timestamp.format(TIME_FORMATTER))
+      .map(StringUtils::urlEncode)
+      .collect(Collectors.joining(",")));
+
+    // TODO REMOVE THIS
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationTimerTask.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      Map<String, ZonedDateTime> request = configs.stream()
+        .collect(Collectors.toMap(NotificationConfiguration::projectKey, NotificationTimerTask::getLastNotificationTime, MERGE_TIMES));
+
+      NotificationChecker notificationChecker = checkerFactory.create(serverConfiguration);
+      List<SonarQubeNotification> notifications = notificationChecker.request(request);
+
+      for (SonarQubeNotification n : notifications) {
+        Stream<NotificationConfiguration> matchingConfStream = configs.stream();
+        if (n.projectKey() != null) {
+          matchingConfStream = matchingConfStream.filter(c -> c.projectKey().equals(n.projectKey()));
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationTimerTask.java]: MAJOR Refactor the code so this stream pipeline is used. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3958)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  SonarQubeNotifications(Timer timer, NotificationTimerTask task) {
+    this.timer = timer;
+    this.task = task;
+    this.timer.scheduleAtFixedRate(task, randomizeInitialDelay(), DELAY);
+  }
+
+  /**
+   * Generated a random number between DELAY*0.5 and DELAY*1.5
+   */
+  private static int randomizeInitialDelay() {
```
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java]: As discussed, I think there isn't much value in this, and in the name of KISS, I'd remove it ;-)

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      return singleton;
+    }
+  }
+
+  /**
+   * Register a project to receive notifications about it.
+   */
+  public void register(NotificationConfiguration configuration) {
+    configuredNotifications.add(configuration);
+    task.setProjects(configuredNotifications);
```
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java]: I see that `setProjects` will make a defensive copy. Why is that necessary? Thanks to `CopyOnWriteArrayList`, it seems the list could be shared.

If the defensive copy is not necessary, and `NotificationTimerTask` can basically share the list directly, then it could be an inner class here, as it's not used anywhere else anyway.

Or, maybe I just overlooked something important ;-)

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * Checks whether a server supports notifications
+   */
+  public boolean isSupported() {
+    String path = getPath(Collections.emptyMap());
+    WsResponse wsResponse = wsClient.rawGet(path);
+    return wsResponse.isSuccessful();
+  }
+
+  private static String getPath(Map<String, ZonedDateTime> requestConfig) {
```
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java]: Hm, sorry, on closer look I see that it's not a URL either. It's path part of a URL + query string. As such, `getUrl` would *not* be better. How about `getWsPath` ?

File: client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/notifications/SonarQubeNotificationTime.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+   * Get the time the last notification was received.
+   * This is the time that will be sent to the next request.
+   * It should be persisted for each project.
+   */
+  ZonedDateTime getLastNotificationTime();
+
+  /**
+   * The date received in notifications will be set here.
+   */
+  void setLastNotificationTime(ZonedDateTime dateTime);
```
Comment (dbmeneses) [inline client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/notifications/SonarQubeNotificationTime.java]: +1

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java (hunk 9 above)
Comment (dbmeneses) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java]: sounds good

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java (hunk 8 above)
Comment (dbmeneses) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java]: Yes, it can be shared. But that implies a sort of contract between both classes (that updates are done through the mutability of the list). 
I think it's nice that `NotificationTimerTask` doesn't get immediately affected by modifications to the list in `SonarQbueNotifications`. It's up to `SonarQubeNotifications` to choose when to explicitly pass a new list to `NotificationTimerTask`.

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java (hunk 7 above)
Comment (dbmeneses) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java]: Ok, I will remove it

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * Get all notification events for a set of projects after a given timestamp. 
+   * Returns an empty list if an error occurred making the request or parsing the response.
+   */
+  @CheckForNull
+  public List<SonarQubeNotification> request(Map<String, ZonedDateTime> projectTimestamps) {
+    String path = getWsPath(projectTimestamps);
+    WsResponse wsResponse = wsClient.rawGet(path);
+    if (!wsResponse.isSuccessful()) {
+      LOG.debug("Failed to get notifications: {}, {}", wsResponse.code(), wsResponse.content());
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/NotificationChecker.java]: MAJOR Invoke method(s) only conditionally. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java (hunk 8 above)
Comment (janos-ss) [inline core/src/main/java/org/sonarsource/sonarlint/core/notifications/SonarQubeNotifications.java]: If `NotificationTimerTask` is an inner class, then it becomes normal for it to be immediately affected.

Anyway, I don't have a strong opinion about this, so it's up to you.
